File: Sorting_Algorithms/Python/mergeSort.py
def merge_sort(arr):
    # Base case: If the array has only one element or is empty, it's already sorted.
    if len(arr) <= 1:
        return arr
    
    # Split the input array into two halves.
    middle = len(arr) // 2
    left_half = arr[:middle]
    right_half = arr[middle:]
    
    # Recursively sort each half.
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)
    
    # Merge the sorted halves back together.
    return merge(left_half, right_half)

def merge(left, right):
    merged = []
    left_index, right_index = 0, 0
    
    # Compare elements from both halves and merge them in sorted order.
    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1
    
    # Append any remaining elements from both halves (if any).
    merged.extend(left[left_index:])
    merged.extend(right[right_index:])
    
    return merged

# Example usage:
arr = [38, 27, 43, 3, 9, 82, 10]
sorted_arr = merge_sort(arr)
print(sorted_arr)


# MERGE SORT

# Divide: First, it takes a big unsorted list and divides it into smaller sub-lists. It keeps doing this until each sub-list contains just one element. At this point, the individual elements are considered sorted by default because there's nothing to compare them to.

# Conquer: Then, it starts merging these smaller sub-lists back together in a sorted manner. It compares the elements in the sub-lists and puts them in the correct order while merging. It does this repeatedly until you end up with one single sorted list.

# The key idea behind merge sort is that it's easier to sort smaller lists and then combine them in a sorted way, rather than trying to sort the entire big list all at once.

# In simpler terms, you can think of it like sorting a deck of cards. First, you divide the deck into smaller piles, then you sort each pile separately (that's the "divide" step), and finally, you merge these sorted piles back together to get one sorted deck (that's the "conquer" step). This process ensures that the deck ends up sorted correctly.

File: C++/mergesort.cpp
#include<iostream>
using namespace std;

void merge(int a[],int left,int mid,int right){
    int n1 = mid - left + 1;
    int n2 = right - mid;
    int leftarray[n1];
    int rightarray[n2];
    for(int i=0;i<n1;i++){
        leftarray[i] = a[left + i];
    }
    for(int j=0;j<n2;j++){
        rightarray[j] = a[mid + 1 + j];
    }
    int i=0,j=0;
    int k = left;
    while(i<n1 && j < n2){
        if(leftarray[i] <= rightarray[j]){
            a[k] = leftarray[i];
            i++;
        }
        else{
            a[k] = rightarray[j];
            j++;

        }
        k++;

    }
    while(i<n1){
        a[k] = leftarray[i];
        i++;
        k++;
    }
    while(j<n2){
        a[k] = rightarray[j];
        j++;
        k++;
    }
}


void mergesort(int a[],int left,int right){
    if(left < right){
        int mid = (left + right)/2;
        mergesort(a,left,mid);
        mergesort(a,mid+1,right);
        merge(a,left,mid,right);
    }
}

int main(){
    int num;
    cout << "Enter the number of elements: " << endl;
    cin >> num;

    int arr[num];
    cout << "Enter the elements:  " << endl;
    for(int i=0;i<num;i++){
        cin >> arr[i];
    }

    cout << "The unsorted array is: " << endl;
    for(int i=0;i<num;i++){
        cout << arr[i] << "  ";
    }cout << endl;

    mergesort(arr,0,num-1);
    cout << "The sorted array is: " << endl;
    for(int i=0;i<num;i++){
        cout << arr[i] << "  ";
    }

    return 0;
}
